helix-core/src/main/java/org/apache/helix/controller/stages/BestPossibleStateCalcStage.java

```diff
@@ -161,49 +162,47 @@ private void addSwapInInstancesToBestPossibleState(Map<String, Resource> resourc
+                // 2. If the swap-out instance's replica is not in the stateMap, set the swap-in instance's replica to the initialState.
+                // This happens when the swap-out node is offline.
+                if (stateMap.containsKey(swapOutInstance)) {
+                  if (stateMap.get(swapOutInstance).equals(stateModelDef.getTopState())
+                      || stateMap.get(swapOutInstance).equals(HelixDefinedState.ERROR.name())) {
+                    // If the swap-out instance's replica is a topState, set the swap-in instance's replica
+                    // to the topState if the StateModel allows another to be added. If not, set the swap-in
+                    // to the secondTopState.
+                    String topStateCount =
+                        stateModelDef.getNumInstancesPerState(stateModelDef.getTopState());
+                    if (topStateCount.equals(
+                        StateModelDefinition.STATE_REPLICA_COUNT_ALL_CANDIDATE_NODES)
+                        || topStateCount.equals(
+                        StateModelDefinition.STATE_REPLICA_COUNT_ALL_REPLICAS)) {
+                      // If the StateModel allows for another replica with the topState to be added,
+                      // set the swap-in instance's replica to the topState.
+                      stateMap.put(swapOutToSwapInInstancePairs.get(swapOutInstance),
+                          stateModelDef.getTopState());
+                    } else {
+                      // If StateModel does not allow another topState replica to be
+                      // added, set the swap-in instance's replica to the secondTopState.
+                      stateMap.put(swapOutToSwapInInstancePairs.get(swapOutInstance),
+                          stateModelDef.getSecondTopStates().iterator().next());
+                    }
                   } else {
-                    // If StateModel does not allow another topState replica to be
-                    // added, set the swap-in instance's replica to the secondTopState.
+                    // If the swap-out instance's replica is not a topState or ERROR, set the swap-in instance's replica
+                    // to the same state
                     stateMap.put(swapOutToSwapInInstancePairs.get(swapOutInstance),
-                        stateModelDef.getSecondTopStates().iterator().next());
+                        stateMap.get(swapOutInstance));
                   }
+                } else {
+                  // If the swap-out instance's replica is not in the stateMap, set the swap-in instance's replica
+                  // to the initialState. This happens when the swap-out node is offline.
+                  stateMap.put(swapOutToSwapInInstancePairs.get(swapOutInstance),
+                      stateModelDef.getInitialState());
                 }
               });
             });
```

Comment:
I would recommend not hybrid the logics here. My understanding of this process is:

1. Get partition mappings to swap-in instances first.
2. Going through again for assigning states with second top (single leader) or top.
3. If you really would like to handle the swap out is already offline case, we can do a minor adjustment. But I personally do not like that way. Because the behavior is unclear to me. Before swap in fully bootstrap and make it as serving node is not the contract we have for this API?